### PR TITLE
 Uniformize kwargs for processors - GroundingDINO

### PR DESCRIPTION
**Pull request by SangbumChoi**
# What does this PR do?

Fixes one model from https://github.com/huggingface/transformers/issues/31911

We don;t want our tracker issue to be closed after the PR is merged, so I edited the magic keyword a bit

## Before submitting
- [] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@zucchini-nlp @NielsRogge @EduardoPach

### CODE REVIEW

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (NielsRogge) [inline src/transformers/models/grounding_dino/processing_grounding_dino.py:113]: FYI the call method now looks uniformly like this:
```
text: Union[TextInput, PreTokenizedInput, List[TextInput], List[PreTokenizedInput]] = None,
audio=None,
videos=None,
**kwargs: Unpack[GroundingDinoProcessorKwargs],
```
I don't see the video keyword argument here so that would need to be added

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (EduardoPach) [inline src/transformers/models/grounding_dino/processing_grounding_dino.py:67]: Should extend the default `ImagesKwargs` in the `image_kwargs` attribute in `ProcessingKwargs` as the default kwargs don't include `annotations`, `do_convert_annotations`, etc. In `GroundingDinoImageProcessor.preprocess`

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (zucchini-nlp) [inline src/transformers/models/grounding_dino/processing_grounding_dino.py:85]: I think we're supposed to add default image kwargs here as well, but I also didn't see default image kwargs on the merged AlignProcessor. @molbap what is the correct way?

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for helping us to make the processors more uniform!

Can you also add a test for this, it is available in:

`from ...test_processing_common import ProcessorTesterMixin`

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:113]: @NielsRogge I think we can just set audio and videos to none in this project but let's see what he says.

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:85]: @zucchini-nlp I was also hesitated to write like this. I will change this as he comments!

Comment (molbap) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:85]: Yes exactly, this should include all defaults for all modalities. However here, the base processor was 
1) not passing any of the kwargs to the `image_processor` 
2) hence, not passing any defaults to the same image processor.
So, we should not pass default image kwargs here, but simply make sure image processing kwargs are received by the image processor post `self._merge_kwargs` as done now.

Comment (molbap) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:85]: ah, just saw @EduardoPach comment above - it seems `annotations` and `do_convert_annotations` should have been passed to the image processor but weren't. So, they don't have currently any default value, the point above does not change

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (molbap) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:113]: I think it's acceptable to set it to None and then to not advertise it in the docstring

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:67]: @EduardoPach I added additional kwargs that used in `GroundingDinoImageProcessor.preprocess` into ImagesKwargs. Does it seems okay?

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py]: @zucchini-nlp @molbap Also now I have noticed that current PR might be having conflict from [PR](https://github.com/huggingface/transformers/pull/31368/files#diff-6505546ec5a9ab74b2ce6511681dd31194eb91e9fa3ce26282e487a5e61f9356). So I will pull his branch whenever it is ready to merge

**Comment by SangbumChoi** [comment]
@zucchini-nlp https://github.com/huggingface/transformers/blob/2c8180acd9d7fc9b91f56425e281ac7237ce3305/tests/test_processing_common.py#L80
I think the circleCI fails since the upper PR is not merged in to main. Otherwise requesting for second review!

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (zucchini-nlp) [inline src/transformers/models/grounding_dino/processing_grounding_dino.py:84]: Hmm, from what i understood in the last message, we don't add defaults here unless they were in processor's call before changes. @molbap to have a look

Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:84]: @zucchini-nlp I am also confused that @molbap https://github.com/huggingface/transformers/blob/2c8180acd9d7fc9b91f56425e281ac7237ce3305/src/transformers/models/bridgetower/processing_bridgetower.py#L43
According to this file we can use some files of images_kwargs perhaps?
 ```
            "do_normalize": True,
            "do_center_crop": True,
```

File: src/transformers/processing_utils.py
Comment (SangbumChoi) [reply src/transformers/processing_utils.py:147]: @zucchini-nlp I can make it to this format. However, I thought the intention of this uniform kwargs was to use single ImagesKwargs to use in every other processor since the variable is can be `Optional`.
Also it is quite difficult to differentiate that these values are specific to DINO-only.

@EduardoPach Any advice for to which variable to move in this `ModelImagesKwargs`?

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (molbap) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:84]: hey @SangbumChoi @zucchini-nlp ! this is made to hold defaults that are already existing in the processor __call__. In this case groundingdino was *not passing any kwarg to the image processor*. So we need to pass image processing args to it, and not add new defaults: in this case, the key `images_kwargs` can be empty.

Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:84]: Now I understood thanks for the clarification!

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:67]: https://github.com/huggingface/transformers/pull/31368#issuecomment-2246659522

I think we can just put the additional value such as `annotations` in the calling processor.

File: src/transformers/processing_utils.py
Comment (zucchini-nlp) [reply src/transformers/processing_utils.py]: General `ImageKwargs` should hold kwargs that are used in almost all models, while `annotations` are used only in some object detection models. Yes, the way we have it now looks good to me

Regarding the tests, we can't merge the PR if tests are failing. So feel free to make changes to the test files to make the CI happy. After that we can request review from core maintainer

Comment (SangbumChoi) [reply src/transformers/processing_utils.py]: @zucchini-nlp Yeah this is still on draft! (I will ping again the PR request from github)

File: src/transformers/processing_utils.py
Comment (SangbumChoi) [inline src/transformers/processing_utils.py:739]: simple fix since it is requires dict(str,str)

**Comment by SangbumChoi** [comment]
@molbap @zucchini-nlp Hi my PR is ready to review. However, since https://github.com/huggingface/transformers/pull/31368 this is ongoing PR, I recommend you to look upper comment. Current CI has error of `align` models which says considered in https://github.com/huggingface/transformers/pull/31368 according to https://github.com/huggingface/transformers/issues/31911. (I think molbap's PR does not currently updated to consider align model)

File: src/transformers/processing_utils.py
Comment (zucchini-nlp) [reply src/transformers/processing_utils.py:801]: Hmm, this is weird because BLIP-2 doesn't have the processor testMixin in the latest `main` branch and thus should not run the failing tests. 

In any case we shouldn't be adding hacks in code to pass the test that shouldn't be running the first place, let's better skip the failing tests in Blip-2 test file with `@unittest.skip("BLIP-2 yet doesn't have standardized API for processor kwargs")`

File: src/transformers/processing_utils.py
Comment (zucchini-nlp) [reply src/transformers/processing_utils.py:829]: This was being handled by the above removed code, if I'm not mistaken. Can you verify that the above code wasn't working properly and add a test for that, to check kwargs are being prioritized properly?

**Review by zucchini-nlp — COMMENTED** [review]
Hey! Thanks for iterating on this. I left comments for the testing code, imo most of it can be removed and is not needed for align, Also, some tests are still failing on CI

File: tests/test_processing_common.py
Comment (SangbumChoi) [reply tests/test_processing_common.py:64]: @zucchini-nlp As I noted this is copied from https://github.com/huggingface/transformers/pull/31368/files#diff-20a0d43edb95664fe346aa1952330dbc88d90b39d06a9c8bf9accfec4ac422b8

Comment (zucchini-nlp) [reply tests/test_processing_common.py:64]: I see, but that PR also has audio processors, while this one doesn't. We usually don't add tests before making required changes to code

Comment (SangbumChoi) [reply tests/test_processing_common.py:64]: @zucchini-nlp Make sense. I just merged molbap's PR since it had some modification that need in here to pass the CI. I will remove the inappropriate commit!

File: src/transformers/processing_utils.py
Comment (SangbumChoi) [reply src/transformers/processing_utils.py:711]: This was also molbap's PR

File: src/transformers/processing_utils.py
Comment (SangbumChoi) [reply src/transformers/processing_utils.py:958]: Also molbap's PR

File: tests/test_processing_common.py
Comment (SangbumChoi) [reply tests/test_processing_common.py:286]: molbap's PR

**Comment by SangbumChoi** [comment]
@zucchini-nlp Sorry for the inconvinience. Merging @molbap ongoing PR made you confuse the overall PR. I will re-comment or fix on the every comment you. Also I will just revert to the main and try to solve as much as I can :)

File: src/transformers/processing_utils.py
Comment (zucchini-nlp) [reply src/transformers/processing_utils.py:711]: Hmm, let's not remove it if it's not affecting your PR and AlignTests

File: src/transformers/processing_utils.py
Comment (molbap) [reply src/transformers/processing_utils.py:958]: there's a discrepancy here - the PR you are using as base had a refactor of the `merge_kwargs` method. I suggest to use the current (merged) version as a basis instead @SangbumChoi !

**Comment by SangbumChoi** [comment]
@zucchini-nlp Hi now I understand the full logic and had a one question on this PR.
It seems that initial PR was designed in here

https://github.com/huggingface/transformers/pull/31197/files#diff-767679f9407c55e91d247ac9ab5359a867414676ae6083959bd2c5cf4e1113c7

with based on following image processor

https://github.com/molbap/transformers/blob/bcce00766b8faa0bd48e0dfe4d0ec72db761f56a/src/transformers/models/efficientnet/image_processing_efficientnet.py

However, not all the processor has the `crop_size` argument for image_processor some of them has only `size`. I think in order to make it model-wise compatibility I recommend to change this to `size`.

https://github.com/molbap/transformers/blob/bcce00766b8faa0bd48e0dfe4d0ec72db761f56a/tests/test_processing_common.py

WDYT? cc. @molbap 

**(BTW, I reverted all the irrelevant commit and major CI fail is related to this comment)**

**Comment by zucchini-nlp** [comment]
IIRC the test will fail if we pass in both, crop_size and size, because one of the args will not be used by image processor. If that's the case, we can override since it is about making tests happy

**Comment by molbap** [comment]
Since only these two arguments are passed, we could also `inspect` in the test method and figure out which is used for a given model, then call with that one - skipping the test entirely is also acceptable, yes

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (SangbumChoi) [inline tests/models/grounding_dino/test_processor_grounding_dino.py:310]: since the padding is default False this should be 4

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (SangbumChoi) [inline tests/models/grounding_dino/test_processor_grounding_dino.py:342]: change all `crop_size` to [size](url)

**Comment by SangbumChoi** [comment]
> IIRC the test will fail if we pass in both, crop_size and size, because one of the args will not be used by image processor. If that's the case, we can override since it is about making tests happy

Well this was not the exactly the case. As I mentioned GroundingDINO processor -> image_processor does not have function of `crop_size`. However, the major test in `test_processor_common.py` uses `crop_size` which fails. I manually changed this to inherit `size` instead of `crop_size`.

@zucchini-nlp Request for the review!

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (zucchini-nlp) [inline tests/models/grounding_dino/test_processor_grounding_dino.py:94]: Seems to be identical to the general `get_component` from Mixin

**Review by zucchini-nlp — APPROVED** [review]
Great work, thanks for working on it! Left one tiny comment

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (molbap) [reply tests/models/grounding_dino/test_processor_grounding_dino.py:94]: +1

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (molbap) [reply tests/models/grounding_dino/test_processor_grounding_dino.py:310]: Note: in that case the test would be moot - the test is to check that kwargs passed (112 max length) do override a default value passed, 117 here. So I recommend passing explicitly `padding="max_length"` here

**Review by molbap — COMMENTED** [review]
Thanks a lot @SangbumChoi :pray: 
I added a comment as well on a test.

File: tests/models/grounding_dino/test_processor_grounding_dino.py
Comment (SangbumChoi) [reply tests/models/grounding_dino/test_processor_grounding_dino.py]: My bad :)

**Review by zucchini-nlp — APPROVED** [review]
Requesting review from core maintainer :)

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (amyeroberts) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py:48]: +1 - we should import here instead

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for handling this!

File: src/transformers/models/grounding_dino/processing_grounding_dino.py
Comment (SangbumChoi) [reply src/transformers/models/grounding_dino/processing_grounding_dino.py]: Done!

**Review by amyeroberts — APPROVED** [review]
❤️

**Pull request merged**
